### ISSUE TITLE
Can no longer zero out the pre-discount field to apply the negative value as a discount after updating to 3.0.6

### ISSUE THREAD

**Issue by tsunamimarketing2** [woocommerce/woocommerce#15104]
### EXPLANATION OF THE ISSUE

After updating to WooCommerce 3.0.6, I can no longer zero out the pre-discount field to apply the negative value as a discount like I could before the update. There was a little issue that I pointed out to you guys about it before that we got used to where we had to enter 0.00 for it to save the change but now that doesn't even work.

Here is a link to a couple tickets I opened about it for more information... 
https://woocommerce.com/my-account/tickets/?id=575105
https://woocommerce.com/my-account/tickets/?id=563994

I already rolled back to 3.0.5 but I would like to be able to keep WooCommerce up to date in the future. So if you could please address this in a future update and let me know when it is fixed, I would greatly appreciate it.

### STEPS TO REPRODUCE THE ISSUE

Either create a new order or edit an existing order... The following screenshot was taken from my site running WooCommerce 3.0.5...
http://tsunamipremiumvapor.com/wp-content/uploads/2017/04/Pre-Discount.png

After updating to 3.0.6 I can no longer enter 0.00 in the pre-discount field to have the negative amount in the "Total" field pulled into the discount field.

This is required for us to be able to sync our orders to QB using Unify because it doesn't let us sync orders that have items with negative amounts.

It has worked the way we need it to in every version WooCommerce up until 3.0.6 and I have already ruled out it being a plugin or a theme conflict in multiple ways. I deactivated all my plugins and switched to the Storefront theme. Then tested it with 3.0.6 and it didn't work. Then I rolled back to 3.0.5 and it worked again.

### SYSTEM STATUS

The following system status is from my site that is running WooCommerce 3.0.5 still...

<details>
`
### WordPress Environment ###

Home URL: https://tsunamipremiumvapor.com
Site URL: https://tsunamipremiumvapor.com
WC Version: 3.0.5
Log Directory Writable: ✔
WP Version: 4.7.4
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache
PHP Version: 7.0.15-0ubuntu0.16.04.4
PHP Post Max Size: 100 MB
PHP Time Limit: 3600
PHP Max Input Vars: 3000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.17
Max Upload Size: 50 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.0.5
WC Database Prefix: wp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔

### Security ###

Secure connection (HTTPS): ❌Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (59) ###

Color and Image Swatches pro for Product Attributes: by Phoeniixx – 1.4
Advanced Database Cleaner: by Younes JFR. – 2.0.0
Advanced Woo Search PRO: by ILLID – 1.00
Autoptimize CriticalCSS.com: by Optimizing Matters – 0.9.2
Apply Coupon on Order for WooCommerce: by CodeWooGeek – 2.2
Authorize.Net CIM for WooCommerce: by Cardpay Solutions
Inc. – 1.0.3

Autoptimize: by Frank Goossens (futtta) – 2.1.0
Awesome Support: E-Mail Support: by The Awesome Support Team – 0.3.0
Awesome Support: by Awesome Support Team – 4.0.0
Better WooCommerce Stars Shortcode: by ClickNathan – 1.0
Disable Comments: by Samir Shah – 1.6
Duplicate Post: by Enrico Battocchi – 3.2
EWWW Image Optimizer Cloud: by Shane Bishop – 3.3.1
Formidable Registration: by Strategy11 – 1.11.09
Formidable Forms: by Strategy11 – 2.03.07
Go - Responsive Pricing & Compare Tables: by Granth – 3.0.1
Complete Analytics Optimization Suite (CAOS): by Daan van den Bergh – 1.44
WPBakery Visual Composer: by Michael M - WPBakery.com – 5.1.1
Lazy Load XT: by Davo Hynds – 0.5.3
Logos Showcase: by Carlos Moreira – 1.7.9
MailChimp for WordPress: by ibericode – 4.1.2
Max Mega Menu - Pro Addon: by Tom Hemsley – 1.5.1
Max Mega Menu: by Tom Hemsley – 2.3.6
Plugins Garbage Collector: by Vladimir Garagulya – 0.10.3
WooCommerce Send Emails: by Stuart O'Brien
cxThemes – 1.3

Temporary Login Without Password: by StoreApps – 1.4
Tesla Framework: by TeslaThemes – 1.2.5
WooCommerce Wishlist Plugin: by TemplateInvaders – 1.3.4
FAQ: by Etoile Web Design – 1.5.34
Uptime Robot Plugin for Wordpress: by Aphotrax – 2.1.2
WC Multiple Email Recipients: by Con Schneider – 1.2.3
Webgility: by Webgility – v346
Woo Email Control: by Ian Young – 1.061
Stamps.com WooCommerce Extension (USPS) (Basic): by XAdapter – 1.1.7
WooCommerce Extended Coupon Features: by Soft79 – 2.4.3
Woocommerce Checkout Terms Conditions Popup: by WC Marketplace
The Grey Parrots – 1.1.7

WooCommerce Expand Tabs: by Diana Burduja – 1.4
WooCommerce Filter Orders by Product: by Md. Kowsar Hossain – 2.0.6
WooCommerce HTML5 Video: by Webilop – 1.7.7
Woocommerce Multiple Addresses: by Alexander Tinyaev – 1.0.7.1
WooCommerce Order Status Manager: by SkyVerge – 1.7.1
WooCommerce Raneri Free Shipping: by Raneri Web Design Snc – 1.47
WooCommerce - ShipStation Integration: by WooCommerce – 4.1.15
WooCommerce Shop as Customer: by cxThemes – 2.09
Woocart Popup Lite for WooCommerce: by Festi-Team – 1.1
WooCommerce: by Automattic – 3.0.5 – 3.0.6 is available
WooCommerce Helper: by WooCommerce – 1.7.2 – Network enabled
WordPress Importer: by wordpressdotorg – 0.6.3
Yoast SEO: by Team Yoast – 4.7.1
WP Add Custom CSS: by Daniele De Santis – 1.0.1
WP Cron Cleaner: by Younes JFR. – 1.0.0
WP Rollback: by WordImpress – 1.4
WP Store Locator: by Tijmen Smit – 2.2.8
WP Video Lightbox: by Tips and Tricks HQ
Ruhul Amin – 1.8.4

WooAge After Sari: by TODCAN Development – 1.00
WPFront User Role Editor Personal Pro: by Syam Mohan – 2.13.1
wpMandrill: by Mandrill – 1.33
YITH WooCommerce Anti-Fraud: by YITHEMES – 1.1.0
YITH WooCommerce Save for Later: by YITHEMES – 1.0.4

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
outofstock (outofstock)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Shop base: #36 - /store/
Cart: #37 - /my-account/cart/
Checkout: #38 - /my-account/checkout/
My account: #39 - /my-account/

### Theme ###

Name: Beatrix Child Theme
Version: 1.0.0
Author URL: http://www.TeslaThemes.com
Child Theme: ✔
Parent Theme Name: Beatrix
Parent Theme Version: 1.0.6
Parent Theme Author URL: http://www.teslathemes.com/
WooCommerce Support: ✔

### Templates ###

Overrides: beatrix-child/woocommerce/archive-product.php
beatrix-child/woocommerce/cart/cart-empty.php
beatrix-child/woocommerce/cart/cart.php
beatrix-child/woocommerce/checkout/form-checkout.php
beatrix-child/woocommerce/content-product.php
beatrix-child/woocommerce/content-single-product.php
beatrix-child/woocommerce/emails/admin-cancelled-order.php
beatrix-child/woocommerce/emails/admin-failed-order.php
beatrix-child/woocommerce/emails/admin-new-order.php
beatrix-child/woocommerce/emails/customer-completed-order.php
beatrix-child/woocommerce/emails/customer-invoice.php
beatrix-child/woocommerce/emails/customer-new-account.php
beatrix-child/woocommerce/emails/customer-on-hold-order.php
beatrix-child/woocommerce/emails/customer-processing-order.php
beatrix-child/woocommerce/emails/customer-refunded-order.php
beatrix-child/woocommerce/emails/customer-reset-password.php
beatrix-child/woocommerce/global/sidebar.php
beatrix-child/woocommerce/global/wrapper-end.php
beatrix-child/woocommerce/global/wrapper-start.php
beatrix-child/woocommerce/loop/loop-end.php
beatrix-child/woocommerce/loop/loop-start.php
beatrix-child/woocommerce/single-product/product-image.php
beatrix-child/woocommerce/single-product/product-thumbnails.php
beatrix-child/woocommerce/single-product.php

### TI WooCommerce Wishlist Templates ###

Overrides: –
`
</details>

![pre-discount](https://cloud.githubusercontent.com/assets/28737020/26119188/67283ae2-3a3a-11e7-96bb-ee2c576b54f1.PNG)

**Comment by mikejolley** [comment]
The change in question formatted values as floats. This means the (correct) validation is kicking in for:

```
// Subtotal cannot be less than total
```

What you're doing is a hack and not truly supported. Even if I were to revert that change, there is no guarantee it would work in future versions.

Why do you need the line item vs just editing the original item's total?

**Comment by tsunamimarketing2** [comment]
I'm not sure what you mean by your first sentence...

What do you mean it is a hack? I didn't edit WooCommerce or anything, this is just how it has always worked.

Because we need to be able to keep track of which damaged products we've replaced. If we just edited the items total, there is no way to keep track of them.

Why did you find it necessary to take away my capability to zero out the pre-discount field could possibly?
 This is really getting on my nerves that you guys refuse to consider fixing this for me since it has always worked like this before 3.0.6. I don't understand why it is such a big deal to revert that change. What will it screw up for people that use your plugin the way they are supposed to?

**Comment by mikejolley** [comment]
By a hack, I mean doing something or taking advantage of an undocumented "feature" or bug.

The reason it will screw up is because our logic etc does not expect the values you're inputting - thats why there is validation code in the first place. Subtotal cannot be less than the total. And line items are not negative.

> Because we need to be able to keep track of which damaged products we've replaced. If we just edited the items total, there is no way to keep track of them.

Yes there is - order notes. A line item isn't easy to track anyway?

In your example you should be editing the line item (the product), leaving subtotal as 69.99, and total as 0. This will then calculate the discount as it should.

**Comment by tsunamimarketing2** [comment]
Order notes do not get imported to QB. We are already mapping the Memo field in QB to import the online order number.

**Comment by mikejolley** [comment]
Does QB import meta data attached to the line item? That would be another place you could make a note.

**Comment by tsunamimarketing2** [comment]
No it doesn't and I can't find any way to get it to.
![unifysettings1](https://cloud.githubusercontent.com/assets/28737020/26124541/d83b2eb2-3a4c-11e7-81cd-f6d40dfd4baf.PNG)
![unifysettings2](https://cloud.githubusercontent.com/assets/28737020/26124543/d9ed42cc-3a4c-11e7-96dc-8a094ccd9e22.PNG)

I attached a couple screenshots of all my options...

**Comment by mikejolley** [comment]
What about the line items? Those are not listed?

Obviously we'll look into this some more but I don't know if anything will change. Validation is important, more important than working around limitations in the QB api :)

**Comment by tsunamimarketing2** [comment]
I'm not sure what you are asking about line items. By line items, do you mean each item in the order?

Each item in the order gets imported to QB but it doesn't include any of the meta data for the items. It just matches each item to the corresponding item in Q.

So it is the QB item that gets inserted into the sales receipts for each order based off the mapping of the product from WooCommerce. We don't need any of the meta data to get imported because all that is already associated with that product in QuickBooks.

So basically, I am stuck with WooCommerce 3.0.5 if you don't find anything that you can change to fix this for me?

**Comment by mikejolley** [comment]
https://github.com/woocommerce/woocommerce/pull/15153 will work around it but it's still an unsupported hack, so you may want to look for another way to send this data in case things change in the future. We do plan on adding a discounts API which could in theory be used for this when built.

**Comment by tsunamimarketing2** [comment]
I tried adding the following to my functions.php file but it didn't fix it...

$this->set_prop( 'total', floatval( wc_format_decimal( $value ) ) );

          // Subtotal cannot be less than total
          if ( '' === $this->get_subtotal() || $this->get_subtotal() < $this->get_total() ) {
              $this->set_subtotal( $value );
          }
      }

Did I do it right?

**Comment by tsunamimarketing2** [comment]
That works, thanks! At-least now I can update WooCommerce as long as I make this change each time until you add the discounts API you were talking about.

Thanks a bunch!

**Comment by tsunamimarketing2** [comment]
Quick question...

If I add the following to my theme's function.php file, will it make it so I don't need to make this change every time I update WooCommerce?

```
// Subtotal cannot be less than total
if ( '' === $this->get_subtotal() || $this->get_subtotal() < $this->get_total() ) {
$this->set_subtotal( $value );
}
```

**Comment by proskibum** [comment]
Just some feedback for this functionality, my client used the pre-discount field to change price for his wholesale customers.  Now it's displaying the discount which he doesn't want, he just wants them to see the unit (changed) price.  It'd be great to get this back in future versions.
